dhall-nixpkgs/Main.hs

```diff
@@ -0,0 +1,948 @@
+
+    exists <- Turtle.testfile expressionFile
+
+    if exists
+        then return ()
+        else do
+            die (MissingFile expressionFile)
+
+    expressionText <- Turtle.readTextFile expressionFile
+
+    let directoryString = Turtle.encodeString directory
+
+    expression <- Dhall.Core.throws (Dhall.Parser.exprFromText directoryString expressionText)
+
+    let status = Dhall.Import.emptyStatus directoryString
+
+    dependencies <- Turtle.reduce Foldl.nub (State.evalStateT (findExternalDependencies expression) status)
+
+    nixDependencies <- traverse dependencyToNix dependencies
+
+    let buildDhallDirectoryPackage = "buildDhallDirectoryPackage"
+
+    let src | null directoryString = directoryString
+            | otherwise            = init directoryString
+
+    let nixExpression =
+            Nix.mkFunction
+                (Nix.mkParamset
+                    (   [ (buildDhallDirectoryPackage, Nothing) ]
+                    <>  nub' (fmap functionParameter nixDependencies)
+                    )
+                    False
+                )
+                (   Nix.mkSym buildDhallDirectoryPackage
+                @@  Nix.attrsE
+                        [ ("name", Nix.mkStr finalName)
+                        , ("src", Nix.mkPath False src)
+                        , ("file", Nix.mkStr (Turtle.format fp file))
+                        , ("source", Nix.mkBool source)
+                        , ("dependencies", Nix.mkList (nub (fmap dependencyExpression nixDependencies)))
+                        ]
+                )
+
+    Prettyprint.Text.putDoc (Nix.Pretty.prettyNix nixExpression)
+
+die :: MonadIO io => Error -> io a
+die e = liftIO $ do
+    Text.IO.hPutStr System.IO.stderr (renderError e)
+
+    System.Exit.exitFailure
+
+data Error
+    = MissingSemanticIntegrityCheck URL
+    | NotAValidGitHubRepositoryURL URL
+    | UnsupportedDomainDependency URL Text
+    | RepositoryIsNotAValidURI Text
+    | UnsupportedURIScheme Text String
+    | UnsupportedDomain Text String
+    | UnsupportedPort Text String
+    | UnsupportedQuery Text String
+    | UnsupportedFragment Text String
+    | NotAValidGitHubRepository Text
+    | FailedToFetchGitHubArchive [Text]
+    | FailedToParseNixPrefetchURL [Text]
+    | FailedToCloneRepository [Text]
+    | FailedToParseNixPrefetchGit [Text] String
+    | MissingFile FilePath
+    
+renderError :: Error -> Text
+renderError e = case e of
+    MissingSemanticIntegrityCheck url ->
+        let dependency = Dhall.Core.pretty url
+
+        in  [NeatInterpolation.text|
+Error: Dependency missing a semantic integrity check
+
+Dhall's Nixpkgs requires that all of your remote dependencies are protected by
+by semantic integrity checks.  This ensures that Nix can replace the remote
+imports with cached imports built by Nix instead of the imports being fetched
+via HTTP requests using Dhall.
+
+The following dependency is missing a semantic integrity check:
+
+↳ $dependency
+|]
```

Comment:
You're not indenting the text to avoid overlong lines, right?

Reply:
@sjakobi: It's mostly because it makes it easier to ensure that the message fits in 80 columns